Feat/mfa channel server

maordadush opened the pull request:
This PR adds MFA authentication support.

widgetii committed:
Artur Zaprzala's rdp2tcp channel driver

widgetii committed:
fix for loop for ancient compilers

widgetii committed:
proper code format

widgetii committed:
VS2010 adaptation

widgetii committed:
format after VS2010

widgetii committed:
remove dead code in generic.c

widgetii committed:
revert entire generic.c

widgetii committed:
revert entire file.c

widgetii committed:
yet another one revertion

kubistika committed:
server/proxy: Fix double-call to disconnect

kubistika committed:
server/proxy: Fix rdpgfx and events leak

kubistika committed:
server/proxy: Fix race-condition

The client's `ClientStop` callback wasn't set, therefore calling
freerdp_client_stop wasn't really blocking until it has stopped, and a
race condition occured. Fixed by registring to `ClientStop` callback and
waiting for the client thread to finish.

kubistika committed:
rdpUpdate: add autoCalculateBitmapData flag

kubistika committed:
server/proxy: Fix bitmap update crash when target is win7

kubistika committed:
rdpgfx/client: Implement CacheImportOffer PDU

kubistika committed:
rdpgfx/client: Expose QoeFrameAck

kubistika committed:
rdpgfx/server: Fix formatting: rdpgfx_recv_cache_import_offer_pdu

kubistika committed:
server/proxy: Proxy CacheImportOffer & QoeFrameAck

kubistika committed:
rdpgfx/client: Update gfx->UnacknowlegedFrame correctly

kubistika committed:
RDPEDISP server side implementation (#5414)

* libfreerdp: Add RDPEDISP server implementation

* server/proxy: Add RDPEDISP support

kubistika committed:
server/proxy: pf_filters: Update filter_free to accept NULL

kubistika committed:
server/proxy: Refactor pdata_set_connection_info

Connection info must be set after the proxy's client connection
established, because if the client didn't provide credentials in .rdp
file, we don't know its username yet.

kubistika committed:
server/proxy: set ps->thread to NULL on disconnect

freerdp_abort_connect calls the proxy's client stop function,
that closes the thread handle. if it already closed by the server,
a segfault will occur because CloseHandle will try to access free()'ed
memory.

kubistika committed:
server/proxy: pf_update: Update ps->settings in DesktopResize

hardening committed:
Merge pull request #5423 from kubistika/feat/bitmap-auto-flag-calculate

rdpUpdate: add autoBitmapDataCalculate

mfleisz committed:
codec: Set alpha channel to FF if not used (#5431)

* Formatting

* codec: Set alpha channel to FF if not used

This PR initializes the alpha channel to FF. Before the alpha channel
contained uninitialized memory potentially causing problems when the
image data was used by an alpha channel aware frontend.

hardening committed:
Merge pull request #5424 from kubistika/rdpgfx

rdpgfx: CacheImportOffer implementation

kubistika committed:
server/proxy: Logging and comments improvements

kubistika committed:
server/proxy: Add check after calling p_client_context_create

akallabeth committed:
Patch from #5436: fix input handling with gstreamer and vaapisink

hardening committed:
Merge pull request #5438 from akallabeth/tsmf_patch

Patch from #5436: fix input handling with gstreamer and vaapisink

akallabeth committed:
Fixed #5441, thank you @ekse

akallabeth committed:
Fixed #5406: Ensure mouse wheel rotation step is never zero

thank you @victorking528

akallabeth committed:
Fix #5427: Dispatch updates to main thread.

thank you @Harrywang55666

akallabeth committed:
Fixed type mismatch.

widgetii committed:
PR fixes

hardening committed:
Merge pull request #5445 from akallabeth/some_small_patches

Fix various small issues

akallabeth committed:
Fix plugin paths when building for a mac bundle.

mfleisz committed:
Merge pull request #5447 from akallabeth/mac_bundle

Fix plugin paths when building for a mac bundle.

akallabeth committed:
Only try to set nego token if load balance info length is >0

mfleisz committed:
Merge pull request #5453 from akallabeth/load_balance_info_checks

Only try to set nego token if load balance info length is >0

kubistika committed:
Declare support for multimon in proxy server.
It was broken in the DISP PR.

mfleisz committed:
rail: Fixes incorrect setting of TS_RAIL_EXEC_FLAG_FILE

The TS_RAIL_EXEC_FLAG should only be set if ExeOrFile refers to a file
path. If we don't supply a published app name (using ||) but the name of
an application (i.e. notepad.exe) this flag was incorrectly set causing
an remote app execution error (error 5 - file not found).

mfleisz committed:
client: populate settings didn't update NetworkAutoDetect

mfleisz committed:
Merge pull request #5458 from kubistika/bugfix/proxy-multimon

Declare support for multimon in proxy server.

mfleisz committed:
Merge pull request #5396 from widgetii/master

Artur Zaprzala's rdp2tcp channel driver

speidy committed:
cmake: add WITH_MFA feature and its dependecies

speidy committed:
channels: mfa: add server-side MFA channel implementation

speidy committed:
server: proxy: implement MFA channel in proxy

kubistika commented on channels/mfa/CMakeLists.txt:
```diff
@@ -0,0 +1,26 @@
+
+define_channel("mfa")
+
+if(WITH_CLIENT_CHANNELS)
```
rm

kubistika commented on channels/mfa/server/CMakeLists.txt:
```diff
@@ -0,0 +1,33 @@
+add_channel_server_library(${MODULE_PREFIX} ${MODULE_NAME} ${CHANNEL_NAME} FALSE "VirtualChannelEntry")
+
+list(APPEND ${MODULE_PREFIX}_LIBS freerdp)
+# list(APPEND ${MODULE_PREFIX}_LIBS winpr)
```
rm

kubistika commented on channels/mfa/server/mfa_main.c:
```diff
@@ -0,0 +1,640 @@
+{
+	UINT error = CHANNEL_RC_OK;
+
+	IFCALLRET(context->ServerInit, error, context);
```
rm callback, no need for this

kubistika commented on include/freerdp/server/mfa.h:
```diff
@@ -0,0 +1,81 @@
+typedef HANDLE (*psMfaGetEventHandle)(MfaServerContext* context);
+typedef UINT (*psMfaCheckEventHandle)(MfaServerContext* context);
+
+typedef UINT (*psMfaServerInit)(MfaServerContext* context);
```
rm

kubistika commented on server/proxy/pf_client.c:
```diff
@@ -50,6 +50,7 @@
 #include "pf_context.h"
 #include "pf_update.h"
 #include "pf_log.h"
+#include "pf_mfa.h"
```
ifdef WITH_MFA

kubistika commented on server/proxy/pf_client.c:
```diff
@@ -246,7 +294,7 @@ static DWORD WINAPI pf_client_thread_proc(LPVOID arg)
 		if (status == WAIT_FAILED)
 		{
-			WLog_ERR(TAG, "%s: WaitForMultipleObjects failed with %"PRIu32"", __FUNCTION__,
+			WLog_ERR(TAG, "%s: WaitForMultipleObjects failed with %" PRIu32 "", __FUNCTION__,
```
fix formatting

kubistika commented on server/proxy/pf_client.c:
```diff
@@ -289,7 +337,7 @@ static int pf_logon_error_info(freerdp* instance, UINT32 data, UINT32 type)
 		return -1;
 
-	pc = (pClientContext*) instance->context;
+	pc = (pClientContext*)instance->context;
```
revert

kubistika commented on server/proxy/pf_context.c:
```diff
@@ -19,6 +19,8 @@
 * limitations under the License.
 */
 
+#include <winpr/synch.h>
```
rm

kubistika commented on server/proxy/pf_context.h:
```diff
@@ -29,10 +33,14 @@
 #include "pf_config.h"
 #include "pf_server.h"
 #include "pf_filters.h"
+#include "pf_mfa.h"
```
ifdef WITH_MFA

kubistika commented on server/proxy/pf_context.h:
```diff
@@ -82,6 +93,10 @@ struct proxy_data
 
 	HANDLE connectionClosed;
+#ifdef WITH_MFA
+	/* used to protect mfaStatus */
```
rm

kubistika commented on server/proxy/pf_mfa.c:
```diff
@@ -0,0 +1,201 @@
+{
+	context->custom = pdata;
+
+	context->ServerInit = pf_mfa_server_init;
```
rm

kubistika commented on server/proxy/pf_server.c:
```diff
@@ -50,6 +54,11 @@
 #include "pf_update.h"
 #include "pf_rdpgfx.h"
 #include "pf_disp.h"
+#include "pf_mfa.h"
```
rm

kubistika replied:
WITH_MFA

kubistika commented on server/proxy/pf_server.c:
```diff
@@ -270,6 +300,7 @@ static DWORD WINAPI pf_server_handle_client(LPVOID arg)
 	client->settings->SuppressOutput = TRUE;
 	client->settings->RefreshRect = TRUE;
 	client->settings->DesktopResize = TRUE;
+
```
rm

kubistika commented on server/proxy/token_validator.h:
```diff
@@ -0,0 +1,51 @@
+struct token_validator;
+typedef struct token_validator TokenValidator;
+
+// TODO: we should decide, right now its arbitrary
```
rm

kubistika commented on server/proxy/token_validator.h:
```diff
@@ -0,0 +1,51 @@
+									 INT64 token_skew_minutes, BOOL insecure_ssl);
+void token_validator_free(TokenValidator* tv);
+BOOL token_validator_validate_token(TokenValidator* tv, proxyData* pdata, const char* compact_token,
+                                    const size_t compact_token_len);
```
formatting?

kubistika replied:
nvm

kubistika commented:
Fixed some problems we've discussed on together. @speidy Please review the changes again and approve

bmiklautz committed:
fix [channels/smartcard]: fix StatusW_Call

According to 2.2.2.18 Status_Call cbAtrLen is unused an must be ignored
upon receipt.

akallabeth committed:
Set freerdp last error on gateway access denied.

nfedera committed:
Merge pull request #5468 from akallabeth/gw_status_denied

Set freerdp last error on gateway access denied.

akallabeth committed:
Don't set ACCESS_DENIED on first HTTP gateway response.

The first response is alwasy ACCESS_DENIED to initiate authentication.
Only the second response must not have that status.

nfedera committed:
Merge pull request #5469 from akallabeth/access_denied

Don't set ACCESS_DENIED on first HTTP gateway response.

kubistika committed:
winpr: ObjectPool_Return: fix wrong type for new_cap

kubistika committed:
winpr: ObjectPool_Return: fix error handling

kubistika committed:
server: proxy: implement NLA fallback

akallabeth committed:
Added PromptForCredentials setting.

akallabeth committed:
Implemented rdp file parser functions with a callback for custom settings.

akallabeth committed:
Fixed rdp file parser tests.

akallabeth committed:
Fixed freerdp_client_rdp_file_set_integer standard flag set.

mfleisz committed:
Merge pull request #5472 from akallabeth/rdp_parser_fixes

Rdp parser fixes

kubistika commented:
@speidy do you think we can merge this one?
rebased on upstream/master

akallabeth committed:
Pass on cert validation failure, set freerdp error in all use cases.

akallabeth committed:
Added freerdp_settings_copy

mfleisz committed:
Merge pull request #5475 from akallabeth/tls_fail_status

Pass on cert validation failure, set freerdp error in all use cases.

kubistika committed:
libfreerdp: certificate: Add key_clone

kubistika committed:
libfreerdp: settings: call key_clone in freerdp_settings_copy

kubistika committed:
libfreerdp: settings: Fix formatting in freerdp_settings_copy

kubistika committed:
server: proxy: use freerdp_settings_copy

kubistika committed:
server: proxy: remove unnecessary files

kubistika committed:
server: proxy: naming refactor

kubistika committed:
server: proxy: set nla fallback flag flag only if NLASecurity is enabled

kubistika committed:
server: proxy: adjust log levels

kubistika committed:
server: proxy: remove unnecessary function handle_client_disconnection

kubistika committed:
server: proxy: Add logs, call client->Close to send disconnection ultimatum

akallabeth committed:
Fix broken clipboard after file copy

akallabeth committed:
Extract whole certificate chain to PEM format.

kubistika committed:
libfreerdp: core: fix const correctness of freerdp_settings_clone

mfleisz committed:
Merge pull request #5481 from akallabeth/pem_intermediate_fix

Extract whole certificate chain to PEM format.

mfleisz committed:
Reformatting

akallabeth committed:
Added alternate full address to rdp file parser.

mfleisz committed:
Merge pull request #5484 from akallabeth/alternate_full_address

Added alternate full address to rdp file parser.

mfleisz committed:
Fix rdp file writer

The RDP file writer was based on previously parsed lines from an
existing rdp file. If you created a new rdpFile, populated it from
settings and tried to write it to a file you just got an error. This PR
fixes this issue by creating the data from the rdpFile properties rather
than the parsed lines.

mfleisz committed:
Merge pull request #5480 from akallabeth/xfreerdp_clip_fix

Fix broken clipboard after file copy

akallabeth committed:
Fixed a bunch of conversion warnings.

kubistika committed:
server: proxy: fix rdpgfx close race

kennylevinsen committed:
uwac: Acknowledge xdg_surface configure messages

This is necessary for the surface to have its role configured.

ligurio committed:
Typo: neigter -> neither

kennylevinsen committed:
uwac: Do not unconditionally set pendingBuffer on configure

If pendingBuffer is set when it was otherwise NULL, then
UwacWindowSubmitBuffer will forever bail early, waiting for a frame
callback that will never come.

kennylevinsen committed:
uwac: Re-enable damage tracking

ligurio committed:
Fixed typo: RAW_VERSTION_STRING -> RAW_VERSION_STRING

kennylevinsen committed:
uwac: Do not memcpy when we swap drawing buffers

The compositor only samples rects that we damage (and have therefore
been overwritten). As we are doing damage tracking and reporting,
memcpy of our old drawing buffer serves no purpose.

kennylevinsen committed:
uwac: Make damage tracking more readable

kennylevinsen committed:
uwac: Only submit dirty buffers

bmiklautz committed:
Merge pull request #5501 from ligurio/typo

Fixed typo: RAW_VERSTION_STRING -> RAW_VERSION_STRING

bmiklautz committed:
Merge pull request #5478 from kubistika/proxy/refactor

Proxy/refactor

bmiklautz committed:
Merge pull request #5498 from ligurio/patch-1

Typo: neigter -> neither

mfleisz committed:
common: Fix parsing of devicestoredirect setting in rdp files

If devicestoredirect was found in an rdp file we incorrectly enabled
drive redirection (which has nothing to do with device redirection).

akallabeth committed:
Fixed loading of sound channel

Ensure sound channel is loaded if settings require it in freerdp_client_load_addins

mfleisz committed:
Merge pull request #5487 from akallabeth/rdp_parser_cleanup

Fixed a bunch of conversion warnings.

akallabeth committed:
With /scale:<value> scale desktop and device.

akallabeth committed:
Fixed loading of microphone channel

Ensure microphone channel is loaded if settings require it in freerdp_client_load_addins

mfleisz committed:
Merge pull request #5505 from akallabeth/rdp_sound

Rdp sound

mfleisz committed:
Merge pull request #5506 from akallabeth/scale_desktop_and_device

With /scale:<value> scale desktop and device.

kubistika committed:
server: proxy: sync server and client channels

when disconnections/redirections occured, sometimes server/client
channels were not synced, meaning that for example the gfx server
received a message, then tried to use gfx client that was already freed.

kubistika committed:
server: proxy: ensure client termination

kubistika committed:
client: xf_channels: remove unused settings variable

kubistika committed:
server: proxy: filters: make sure filter_init is exported in filters compilation

kubistika committed:
server: proxy: filters: demo: log instead of dropping the connection

kubistika committed:
server: proxy: channels: only open allowed channels

kubistika committed:
server: proxy: refactor configuration loader

kubistika committed:
server: proxy: print configuration on startup

hardening committed:
Merge pull request #5511 from kubistika/xf-unused-var

client: xf_channels: remove unused settings variable

hardening committed:
Merge pull request #5512 from kubistika/proxy/config-loader-refactor

Proxy/config loader refactor
